File: plugins/diagnostics-panel.lua
-- Diagnostics Panel Plugin
-- Displays LSP diagnostics in a virtual buffer with Emacs-style navigation
--
-- Features:
-- - Opens in a horizontal split below current buffer
-- - Navigate with Up/Down arrows or n/p keys
-- - Jump to diagnostic location with Enter

-- Track the diagnostics panel state
local panel_state = {
    open = false,
    current_index = 1,  -- Currently selected diagnostic (1-indexed)
    diagnostics = {},   -- Current diagnostics data
    header_lines = 2,   -- Number of header lines before first diagnostic
}

-- Initialize the diagnostics-list mode
local function setup_mode()
    editor.define_mode({
        name = "diagnostics-list",
        parent = "special",  -- Inherits 'q' to quit, 'g' to refresh
        bindings = {
            ["RET"] = "diagnostics:goto",      -- Jump to diagnostic location
            ["n"] = "diagnostics:next",        -- Move to next diagnostic
            ["p"] = "diagnostics:prev",        -- Move to previous diagnostic
            ["Down"] = "diagnostics:next",     -- Arrow key support
            ["Up"] = "diagnostics:prev",       -- Arrow key support
            ["j"] = "diagnostics:next",        -- Vim-style
            ["k"] = "diagnostics:prev",        -- Vim-style
        },
        read_only = true
    })
    debug("Registered diagnostics-list mode")
end

-- Format a diagnostic for display
local function format_diagnostic(diag, index, is_selected)
    local severity_icon = {
        error = "E",
        warning = "W",
        info = "I",
        hint = "H"
    }

    local icon = severity_icon[diag.severity] or "?"
    local marker = is_selected and ">" or " "
    return string.format("%s [%s] %s:%d:%d: %s\n",
        marker, icon, diag.file, diag.line, diag.column, diag.message)
end

-- Build entries from diagnostics data
local function build_entries()
    local entries = {}

    -- Add header
    table.insert(entries, {
        text = "=== LSP Diagnostics ===\n\n",
        properties = {}
    })

    -- Add each diagnostic with embedded properties
    for i, diag in ipairs(panel_state.diagnostics) do
        local is_selected = (i == panel_state.current_index)
        local text = format_diagnostic(diag, i, is_selected)
        table.insert(entries, {
            text = text,
            properties = {
                diagnostic_index = i,
                file = diag.file,
                line = diag.line,
                column = diag.column,
                severity = diag.severity,
                message = diag.message
            }
        })
    end

    -- Add footer
    table.insert(entries, {
        text = string.format("\nTotal: %d diagnostics (selected: %d/%d)",
            #panel_state.diagnostics, panel_state.current_index, #panel_state.diagnostics),
        properties = {}
    })

    return entries
end

-- Generate mock diagnostics (in a real implementation, these would come from LSP)
local function get_diagnostics()
    return {
        {
            severity = "error",
            file = "src/main.rs",
            line = 42,
            column = 5,
            message = "mismatched types: expected `usize`, found `i32`"
        },
        {
            severity = "warning",
            file = "src/lib.rs",
            line = 156,
            column = 1,
            message = "unused variable: `temp`"
        },
        {
            severity = "error",
            file = "src/editor.rs",
            line = 1024,
            column = 12,
            message = "cannot borrow `*self` as mutable because it is also borrowed as immutable"
        },
        {
            severity = "info",
            file = "src/plugin_api.rs",
            line = 89,
            column = 8,
            message = "consider using `&str` instead of `String`"
        },
        {
            severity = "hint",
            file = "src/buffer.rs",
            line = 203,
            column = 15,
            message = "this function has too many arguments (6/5)"
        }
    }
end

-- Create or update the diagnostic panel
local function show_panel()
    -- Load diagnostics
    panel_state.diagnostics = get_diagnostics()
    panel_state.current_index = 1  -- Reset selection to first diagnostic

    if #panel_state.diagnostics == 0 then
        editor.set_status("No diagnostics to display")
        return
    end

    -- Build entries with text properties
    local entries = build_entries()

    debug(string.format("Creating diagnostics panel with %d entries", #entries))

    -- Create the virtual buffer in a horizontal split
    -- The ratio 0.7 means top pane (code) gets 70%, bottom pane (diagnostics) gets 30%
    -- panel_id makes this idempotent - if panel exists, just update its content
    editor.create_virtual_buffer_in_split({
        name = "*Diagnostics*",
        mode = "diagnostics-list",
        read_only = true,
        entries = entries,
        ratio = 0.7,
        panel_id = "diagnostics"  -- Unique ID for idempotent updates
    })

    panel_state.open = true
    editor.set_status(string.format("Diagnostics: %d items - use Up/Down to navigate, Enter to jump",
        #panel_state.diagnostics))
end

-- Toggle the diagnostics panel
function toggle_diagnostics_panel()
    debug("Toggling diagnostics panel")
    show_panel()
end

-- Jump to the diagnostic at the current selection
function goto_diagnostic()
    local diag = panel_state.diagnostics[panel_state.current_index]
    if not diag then
        editor.set_status("No diagnostic selected")
        return
    end

    debug(string.format("Jumping to %s:%d:%d", diag.file, diag.line, diag.column))

    -- Open the file at the diagnostic location
    editor.open_file_at_location(diag.file, diag.line, diag.column)
    editor.set_status(string.format("Jumped to %s:%d:%d - %s",
        diag.file, diag.line, diag.column, diag.message))
end

-- Move to next diagnostic
function diagnostics_next()
    if #panel_state.diagnostics == 0 then
        editor.set_status("No diagnostics")
        return
    end

    panel_state.current_index = panel_state.current_index + 1
    if panel_state.current_index > #panel_state.diagnostics then
        panel_state.current_index = 1  -- Wrap around
    end

    local diag = panel_state.diagnostics[panel_state.current_index]
    editor.set_status(string.format("[%d/%d] %s:%d - %s",
        panel_state.current_index, #panel_state.diagnostics,
        diag.file, diag.line, diag.message))

    -- Rebuild the panel to show the new selection
    local entries = build_entries()
    -- Note: We can't update content of existing buffer yet,
    -- so we recreate the panel (this is a limitation)
    debug(string.format("Selected diagnostic %d", panel_state.current_index))
end

-- Move to previous diagnostic
function diagnostics_prev()
    if #panel_state.diagnostics == 0 then
        editor.set_status("No diagnostics")
        return
    end

    panel_state.current_index = panel_state.current_index - 1
    if panel_state.current_index < 1 then
        panel_state.current_index = #panel_state.diagnostics  -- Wrap around
    end

    local diag = panel_state.diagnostics[panel_state.current_index]
    editor.set_status(string.format("[%d/%d] %s:%d - %s",
        panel_state.current_index, #panel_state.diagnostics,
        diag.file, diag.line, diag.message))

    -- Rebuild the panel to show the new selection
    local entries = build_entries()
    -- Note: We can't update content of existing buffer yet,
    -- so we recreate the panel (this is a limitation)
    debug(string.format("Selected diagnostic %d", panel_state.current_index))
end

-- Register commands
editor.register_command({
    name = "Show Diagnostics",
    description = "Show LSP diagnostics in a panel",
    action = "toggle_diagnostics_panel",
    contexts = {"normal"}
})

editor.register_command({
    name = "Goto Diagnostic",
    description = "Jump to diagnostic source location",
    action = "goto_diagnostic",
    contexts = {"normal"}
})

editor.register_command({
    name = "Next Diagnostic",
    description = "Move to next diagnostic",
    action = "diagnostics_next",
    contexts = {"normal"}
})

editor.register_command({
    name = "Previous Diagnostic",
    description = "Move to previous diagnostic",
    action = "diagnostics_prev",
    contexts = {"normal"}
})

-- Initialize the mode on load
setup_mode()

debug("Diagnostics panel plugin loaded")
editor.set_status("Diagnostics panel plugin ready - use 'Show Diagnostics' command")

File: plugins/diagnostics_panel.ts
/// <reference path="./lib/fresh.d.ts" />

/**
 * Diagnostics Panel Plugin
 *
 * Interactive diagnostics panel showing LSP diagnostics with:
 * - Real-time updates when diagnostics change
 * - Filter by current file or show all files
 * - Cursor navigation with highlighting
 * - Enter to jump to diagnostic location
 */

// =============================================================================
// Types and Interfaces
// =============================================================================

interface DiagnosticLocation {
  file: string;
  line: number;
  column: number;
}

interface DiagnosticLineMapping {
  panelLine: number;  // 1-based line in panel
  location: DiagnosticLocation;
}

interface DiagnosticsState {
  isOpen: boolean;
  bufferId: number | null;
  splitId: number | null;
  sourceSplitId: number | null;  // The split that was active when panel opened
  sourceBufferId: number | null;
  showAllFiles: boolean;
  cachedContent: string;
  // Maps panel line numbers to diagnostic locations for sync
  lineMappings: DiagnosticLineMapping[];
  // Whether the panel itself is focused (freezes source buffer tracking)
  panelFocused: boolean;
  // Current cursor line in the panel (1-indexed)
  panelCursorLine: number;
}

// =============================================================================
// State Management
// =============================================================================

const state: DiagnosticsState = {
  isOpen: false,
  bufferId: null,
  splitId: null,
  sourceSplitId: null,
  sourceBufferId: null,
  showAllFiles: false,  // Default to filtering by current file
  cachedContent: "",
  lineMappings: [],
  panelFocused: false,
  panelCursorLine: 1,
};

// =============================================================================
// Color Definitions
// =============================================================================

const colors = {
  error: [255, 100, 100] as [number, number, number],
  warning: [255, 200, 100] as [number, number, number],
  info: [100, 200, 255] as [number, number, number],
  hint: [150, 150, 150] as [number, number, number],
  file: [180, 180, 255] as [number, number, number],
  location: [150, 255, 150] as [number, number, number],
  header: [255, 200, 100] as [number, number, number],
  selected: [80, 80, 120] as [number, number, number],
};

// =============================================================================
// Mode Definition
// =============================================================================

editor.defineMode(
  "diagnostics-list",
  "normal",
  [
    ["Return", "diagnostics_goto"],
    ["Tab", "diagnostics_goto"],
    ["a", "diagnostics_toggle_all"],
    ["r", "diagnostics_refresh"],
    ["q", "diagnostics_close"],
    ["Escape", "diagnostics_close"],
  ],
  true
);

// =============================================================================
// Helpers
// =============================================================================

function severityIcon(severity: number): string {
  switch (severity) {
    case 1: return "[E]";
    case 2: return "[W]";
    case 3: return "[I]";
    case 4: return "[H]";
    default: return "[?]";
  }
}

function uriToPath(uri: string): string {
  if (uri.startsWith("file://")) {
    return uri.slice(7);
  }
  return uri;
}

function getActiveFileUri(): string | null {
  const bufferId = state.sourceBufferId ?? editor.getActiveBufferId();
  const path = editor.getBufferPath(bufferId);
  if (!path) return null;
  return "file://" + path;
}

function entriesToContent(entries: TextPropertyEntry[]): string {
  return entries.map(e => e.text).join("");
}

// =============================================================================
// Panel Content Building
// =============================================================================

function buildPanelEntries(): TextPropertyEntry[] {
  const entries: TextPropertyEntry[] = [];
  const diagnostics = editor.getAllDiagnostics();

  // Clear and rebuild line mappings
  state.lineMappings = [];

  const activeUri = getActiveFileUri();
  const filterUri = state.showAllFiles ? null : activeUri;

  // Filter diagnostics
  const filtered = filterUri
    ? diagnostics.filter(d => d.uri === filterUri)
    : diagnostics;

  // Group by file
  const byFile = new Map<string, TsDiagnostic[]>();
  for (const diag of filtered) {
    const existing = byFile.get(diag.uri) || [];
    existing.push(diag);
    byFile.set(diag.uri, existing);
  }

  // Sort files, with active file first if filtering
  const files = Array.from(byFile.keys()).sort((a, b) => {
    if (activeUri) {
      if (a === activeUri) return -1;
      if (b === activeUri) return 1;
    }
    // Simple string comparison (localeCompare has ICU issues in Deno)
    if (a < b) return -1;
    if (a > b) return 1;
    return 0;
  });

  // Header (line 1)
  let filterLabel: string;
  if (state.showAllFiles) {
    filterLabel = "All Files";
  } else if (activeUri) {
    const fileName = editor.pathBasename(uriToPath(activeUri));
    filterLabel = fileName;
  } else {
    filterLabel = "Current File";
  }
  entries.push({
    text: `Diagnostics (${filterLabel}):\n`,
    properties: { type: "header" },
  });

  let currentPanelLine = 2; // Start after header

  if (filtered.length === 0) {
    entries.push({
      text: "  No diagnostics\n",
      properties: { type: "empty" },
    });
    currentPanelLine++;
  } else {
    let diagIndex = 0;
    for (const uri of files) {
      const fileDiags = byFile.get(uri) || [];
      const filePath = uriToPath(uri);
      const fileName = editor.pathBasename(filePath);

      // File header (blank line + filename)
      entries.push({
        text: `\n${fileName}:\n`,
        properties: { type: "file-header", uri },
      });
      currentPanelLine += 2; // blank line + file header

      // Sort diagnostics by line, then severity
      fileDiags.sort((a, b) => {
        const lineDiff = a.range.start.line - b.range.start.line;
        if (lineDiff !== 0) return lineDiff;
        return a.severity - b.severity;
      });

      for (const diag of fileDiags) {
        const icon = severityIcon(diag.severity);
        const line = diag.range.start.line + 1;
        const col = diag.range.start.character + 1;
        const msg = diag.message.split("\n")[0]; // First line only

        const location: DiagnosticLocation = {
          file: filePath,
          line: line,
          column: col,
        };

        // Track mapping for cursor sync
        state.lineMappings.push({
          panelLine: currentPanelLine,
          location: location,
        });

        entries.push({
          text: `  ${icon} ${line}:${col} ${msg}\n`,
          properties: {
            type: "diagnostic",
            index: diagIndex,
            severity: diag.severity,
            location: location,
          },
        });
        diagIndex++;
        currentPanelLine++;
      }
    }
  }

  // Summary
  const errorCount = filtered.filter(d => d.severity === 1).length;
  const warningCount = filtered.filter(d => d.severity === 2).length;
  const infoCount = filtered.filter(d => d.severity === 3).length;

  entries.push({
    text: "\n",
    properties: { type: "blank" },
  });
  entries.push({
    text: `${errorCount}E ${warningCount}W ${infoCount}I | a: toggle filter | r: refresh | RET: goto | q: close\n`,
    properties: { type: "footer" },
  });

  return entries;
}

// =============================================================================
// Highlighting
// =============================================================================

function applyHighlighting(): void {
  if (state.bufferId === null) return;

  const bufferId = state.bufferId;
  editor.clearNamespace(bufferId, "diag");

  const content = state.cachedContent;
  if (!content) return;

  const lines = content.split("\n");
  const cursorLine = state.panelCursorLine;

  let byteOffset = 0;

  for (let lineIdx = 0; lineIdx < lines.length; lineIdx++) {
    const line = lines[lineIdx];
    const lineStart = byteOffset;
    const lineEnd = byteOffset + line.length;
    const isCurrentLine = (lineIdx + 1) === cursorLine;
    const isDiagnosticLine = line.trim().startsWith("[");

    // Highlight current line if it's a diagnostic line (entire line gets background)
    if (isCurrentLine && isDiagnosticLine) {
      editor.addOverlay(
        bufferId, "diag", lineStart, lineEnd,
        colors.selected[0], colors.selected[1], colors.selected[2],
        true, true, false
      );
    }

    // Header highlighting
    if (line.startsWith("Diagnostics")) {
      editor.addOverlay(
        bufferId, "diag", lineStart, lineEnd,
        colors.header[0], colors.header[1], colors.header[2],
        true, true, false
      );
    }

    // File header highlighting
    if (line.endsWith(":") && !line.startsWith("Diagnostics") && !line.startsWith(" ")) {
      editor.addOverlay(
        bufferId, "diag", lineStart, lineEnd,
        colors.file[0], colors.file[1], colors.file[2],
        false, true, false
      );
    }

    // Severity icon highlighting
    const iconMatch = line.match(/^\s+\[([EWIH?])\]/);
    if (iconMatch) {
      const iconStart = lineStart + line.indexOf("[");
      const iconEnd = iconStart + 3;

      let color: [number, number, number];
      switch (iconMatch[1]) {
        case "E": color = colors.error; break;
        case "W": color = colors.warning; break;
        case "I": color = colors.info; break;
        case "H": color = colors.hint; break;
        default: color = colors.hint;
      }

      editor.addOverlay(
        bufferId, "diag", iconStart, iconEnd,
        color[0], color[1], color[2],
        false, true, false
      );

      // Location highlighting (line:col after icon)
      const locMatch = line.match(/\[.\]\s+(\d+:\d+)/);
      if (locMatch && locMatch.index !== undefined) {
        const locStart = lineStart + line.indexOf(locMatch[1]);
        const locEnd = locStart + locMatch[1].length;
        editor.addOverlay(
          bufferId, "diag", locStart, locEnd,
          colors.location[0], colors.location[1], colors.location[2],
          false, false, false
        );
      }
    }

    byteOffset += line.length + 1;
  }
}

function updatePanel(): void {
  if (state.bufferId === null) return;

  const entries = buildPanelEntries();
  state.cachedContent = entriesToContent(entries);
  editor.setVirtualBufferContent(state.bufferId, entries);
  applyHighlighting();
}

// =============================================================================
// Commands
// =============================================================================

globalThis.show_diagnostics_panel = async function(): Promise<void> {
  if (state.isOpen) {
    // If already open, just focus the panel
    if (state.splitId !== null) {
      editor.focusSplit(state.splitId);
    }
    return;
  }

  state.sourceSplitId = editor.getActiveSplitId();
  state.sourceBufferId = editor.getActiveBufferId();
  state.panelFocused = false;

  const entries = buildPanelEntries();
  state.cachedContent = entriesToContent(entries);

  // Create a horizontal split below the current buffer
  const result = await editor.createVirtualBufferInSplit({
    name: "*Diagnostics*",
    mode: "diagnostics-list",
    read_only: true,
    entries: entries,
    ratio: 0.7,  // Source keeps 70%, panel takes 30%
    direction: "horizontal",  // Split below
    panel_id: "diagnostics",  // Enable idempotent updates
    show_line_numbers: false,
    show_cursors: true,
    editing_disabled: true,
  });

  if (result.buffer_id !== null) {
    state.isOpen = true;
    state.bufferId = result.buffer_id;
    state.splitId = result.split_id ?? null;
    state.panelFocused = true;  // Panel starts focused
    applyHighlighting();

    const diagnostics = editor.getAllDiagnostics();
    editor.setStatus(`Diagnostics: ${diagnostics.length} items | a: toggle filter | RET: goto | q: close`);
  } else {
    state.sourceSplitId = null;
    state.sourceBufferId = null;
    editor.setStatus("Failed to open diagnostics panel");
  }
};

globalThis.diagnostics_close = function(): void {
  if (!state.isOpen) return;

  // Capture values before clearing state
  const splitId = state.splitId;
  const sourceSplitId = state.sourceSplitId;
  const sourceBufferId = state.sourceBufferId;
  const bufferId = state.bufferId;

  // Clear state FIRST to prevent event handlers from trying to update
  state.isOpen = false;
  state.bufferId = null;
  state.splitId = null;
  state.sourceSplitId = null;
  state.sourceBufferId = null;
  state.cachedContent = "";
  state.panelFocused = false;

  // Try to close the split first
  let splitClosed = false;
  if (splitId !== null) {
    splitClosed = editor.closeSplit(splitId);
  }

  // If split couldn't be closed (only split), switch buffer back to source
  if (!splitClosed && splitId !== null && sourceBufferId !== null) {
    editor.setSplitBuffer(splitId, sourceBufferId);
  }

  // Always delete the virtual buffer completely (removes from all splits)
  if (bufferId !== null) {
    editor.closeBuffer(bufferId);
  }

  // Focus back on the source split
  if (sourceSplitId !== null) {
    editor.focusSplit(sourceSplitId);
  }

  editor.setStatus("Diagnostics panel closed");
};

globalThis.diagnostics_goto = function(): void {
  if (!state.isOpen || state.bufferId === null) return;

  const props = editor.getTextPropertiesAtCursor(state.bufferId);

  if (props.length > 0) {
    const location = props[0].location as { file: string; line: number; column: number } | undefined;
    if (location) {
      // Close panel first, then open file
      const file = location.file;
      const line = location.line;
      const col = location.column;

      globalThis.diagnostics_close();
      editor.openFile(file, line, col);
      editor.setStatus(`Jumped to ${editor.pathBasename(file)}:${line}`);
      return;
    }
  }

  editor.setStatus("Move cursor to a diagnostic line");
};

globalThis.diagnostics_toggle_all = function(): void {
  if (!state.isOpen) return;

  state.showAllFiles = !state.showAllFiles;
  updatePanel();

  const label = state.showAllFiles ? "All Files" : "Current File";
  editor.setStatus(`Showing: ${label}`);
};

globalThis.diagnostics_refresh = function(): void {
  if (!state.isOpen) return;

  updatePanel();
  editor.setStatus("Diagnostics refreshed");
};

globalThis.toggle_diagnostics_panel = function(): void {
  if (state.isOpen) {
    globalThis.diagnostics_close();
  } else {
    globalThis.show_diagnostics_panel();
  }
};

// =============================================================================
// Event Handlers
// =============================================================================

// Find the panel line that matches a source file location
function findPanelLineForLocation(file: string, sourceLine: number): number | null {
  // Find the first diagnostic on this source line for this file
  for (const mapping of state.lineMappings) {
    if (mapping.location.file === file && mapping.location.line === sourceLine) {
      return mapping.panelLine;
    }
  }
  return null;
}

// Convert a 1-based line number to byte offset in the cached content
function lineToByteOffset(lineNumber: number): number {
  const lines = state.cachedContent.split("\n");
  let offset = 0;
  for (let i = 0; i < lineNumber - 1 && i < lines.length; i++) {
    offset += lines[i].length + 1; // +1 for newline
  }
  return offset;
}

// Sync the panel cursor to match a source location
function syncPanelCursorToSourceLine(file: string, sourceLine: number): void {
  if (state.bufferId === null) return;

  const panelLine = findPanelLineForLocation(file, sourceLine);
  if (panelLine !== null) {
    // Convert panel line number to byte offset and move cursor
    const byteOffset = lineToByteOffset(panelLine);
    state.panelCursorLine = panelLine;
    editor.setBufferCursor(state.bufferId, byteOffset);
    applyHighlighting();
  }
}

globalThis.on_diagnostics_cursor_moved = function(data: {
  buffer_id: number;
  cursor_id: number;
  old_position: number;
  new_position: number;
  line: number;
}): void {
  if (!state.isOpen || state.bufferId === null) return;

  // If cursor moved in the diagnostics panel, update the tracked line and highlighting
  if (data.buffer_id === state.bufferId) {
    state.panelCursorLine = data.line;
    applyHighlighting();
    return;
  }

  // If cursor moved in the source buffer and panel is not focused, sync the panel cursor
  if (data.buffer_id === state.sourceBufferId && !state.panelFocused) {
    const path = editor.getBufferPath(data.buffer_id);
    if (path) {
      // Use the line number from the hook (1-indexed)
      syncPanelCursorToSourceLine(path, data.line);
    }
  }
};

globalThis.on_diagnostics_updated = function(_data: {
  uri: string;
  count: number;
}): void {
  if (!state.isOpen) return;
  updatePanel();
};

globalThis.on_diagnostics_buffer_activated = function(data: {
  buffer_id: number;
}): void {
  if (!state.isOpen) return;

  // If the diagnostics panel became active, mark as focused (freeze source buffer)
  if (data.buffer_id === state.bufferId) {
    state.panelFocused = true;
    return;
  }

  // If we're focusing a different buffer and were in the panel, mark as unfocused
  if (state.panelFocused) {
    state.panelFocused = false;
  }

  // Update source buffer and refresh the panel to show diagnostics for the new buffer
  state.sourceBufferId = data.buffer_id;
  updatePanel();
};

// Register event handlers
editor.on("cursor_moved", "on_diagnostics_cursor_moved");
editor.on("diagnostics_updated", "on_diagnostics_updated");
editor.on("buffer_activated", "on_diagnostics_buffer_activated");

// =============================================================================
// Command Registration
// =============================================================================

editor.registerCommand(
  "Show Diagnostics Panel",
  "Open the diagnostics panel",
  "show_diagnostics_panel",
  "normal"
);

editor.registerCommand(
  "Toggle Diagnostics Panel",
  "Toggle the diagnostics panel",
  "toggle_diagnostics_panel",
  "normal"
);

// =============================================================================
// Initialization
// =============================================================================

editor.setStatus("Diagnostics Panel plugin loaded");
editor.debug("Diagnostics Panel plugin initialized");
